### PR TITLE
[PRM-1008] Create Controller/Parser + Log message

### PR DESCRIPTION
**Pull request by ruthdas03**

### CODE REVIEW

File: test/controllers/SDESCallbackControllerSpec.scala [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+       |            ]
+       |        }
+       |""".stripMargin
+  )
+
+  "receiveSDESCallback" should {
+    "return content string for Valid sdesCallback JSON Body" when {
+      "the JSON request body is valid" in new Setup {
+        val result: Future[Result] = sdesController.handleCallback()(fakeRequest.withJsonBody(sdesCallbackJson))
+         contentAsString(result) shouldBe "Valid SDESCallback body"
```
Comment (LewisMitchell) [inline test/controllers/SDESCallbackControllerSpec.scala:70]: I think a status check would be sufficient here rather than checking the body

**Review by LewisMitchell — CHANGES_REQUESTED** [review]
Looks good, just some changes

File: app/controllers/SDESCallbackController.scala [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+            failure => {
+              logger.error("[SDESCallbackController][handleCallback] Fail to parse request body to model")
+              logger.debug(s"[SDESCallbackController][handleCallback] Parse failure(s): $failure")
+              println("Failure to parse..."+ failure)
+              Future(BadRequest("Failed to parse to model"))
+            },
+            sdesCallback => {
+              logger.info(s"[SDESCallbackController][handleCallback] Callback received from SDES with status ${sdesCallback.notification}. SDESCallback received  = $sdesCallback.toString")
+              auditService.audit(FileNotificationStatusAuditModel(sdesCallback.notification, sdesCallback.filename, sdesCallback.correlationID, sdesCallback.failureReason, sdesCallback.availableUntil))
+              Future(Ok("Valid SDESCallback body"))
```
[the remark this answers is not in the document]
Comment (AndrewRButcher) [reply app/controllers/SDESCallbackController.scala:51]: The  SDES API spec is expecting it to return 204 No Content as Lewis has suggested
Below is the responses SDES are expecting 

Code | Description
-- | --
204 | No Content
400 | Bad Request. Invalid Json. Parameter {name} is missing or invalid
500 | Internal Server Error

File: app/controllers/SDESCallbackController.scala [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        logger.error("[SDESCallbackController][handleCallback] Failed to validate request body as JSON")
+        Future(BadRequest("Invalid body received i.e. could not be parsed to JSON"))
+      })(
+        jsonBody => {
+          val parseResultToModel = Json.fromJson(jsonBody)(SDESCallback.apiSDESCallbackReads)
+          parseResultToModel.fold(
+            failure => {
+              logger.error("[SDESCallbackController][handleCallback] Fail to parse request body to model")
+              logger.debug(s"[SDESCallbackController][handleCallback] Parse failure(s): $failure")
+              println("Failure to parse..."+ failure)
```
[the remark this answers is not in the document]
Comment (ruthdas03) [reply app/controllers/SDESCallbackController.scala]: Removed println, thanks.

File: app/controllers/SDESCallbackController.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (ruthdas03) [reply app/controllers/SDESCallbackController.scala]: Thanks, replaced Ok with No_Content.

File: app/models/SDESCallback.scala [hunk 4]
```diff
@@ -35,7 +35,26 @@ case class SDESCallback(
...
+      dateTime <- (json \ "dateTime").validate[LocalDateTime]
+      properties <-
+        (json \ "properties").validate[Properties](
+          Properties.formatProperties
+        )
+    } yield {
+      SDESCallback(notification, filename, Some(checksumAlgorithm), Some(checksum), correlationID,
+          Some(availableUntil), Some(failureReason), Some(dateTime), Some(Seq(properties)))
+    }
+  }
```
[the remark this answers is not in the document]
Comment (ruthdas03) [reply app/models/SDESCallback.scala]: Agreed, hence used the suggested code.

File: app/models/SDESCallback.scala [hunk 5]
```diff
@@ -35,7 +35,26 @@ case class SDESCallback(
...
+      availableUntil <- (json \ "availableUntil").validate[LocalDateTime]
+      failureReason <- (json \ "failureReason").validate[String]
+      dateTime <- (json \ "dateTime").validate[LocalDateTime]
+      properties <-
+        (json \ "properties").validate[Properties](
+          Properties.formatProperties
+        )
+    } yield {
+      SDESCallback(notification, filename, Some(checksumAlgorithm), Some(checksum), correlationID,
+          Some(availableUntil), Some(failureReason), Some(dateTime), Some(Seq(properties)))
```
[the remark this answers is not in the document]
Comment (ruthdas03) [reply app/models/SDESCallback.scala]: Well not keeping this and used the above changes instead.

File: it/controllers/SDESCallbackControllerISpec.scala [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+       |                    "name": "name1",
+       |                    "value": "value1"
+       |                }
+       |            ]
+       |        }
+       |""".stripMargin
+  )
+
+  "handleCallback" should {
+    "return OK (200)" when {
```
[the remark this answers is not in the document]
Comment (ruthdas03) [reply it/controllers/SDESCallbackControllerISpec.scala]: Agreed and Changed.

File: test/controllers/SDESCallbackControllerSpec.scala [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+import scala.concurrent.Future
+
+class SDESCallbackControllerSpec extends SpecBase {
+  val mockService: AuditService = mock(classOf[AuditService])
+
+  class Setup() {
+    val sdesController = new SDESCallbackController(mockService, stubControllerComponents())
+  }
+
+  val sdesCallbackModel: SDESCallback = SDESCallback(SDESFileNotificationEnum.FileReady,"axyz.doc",Some("MD5"),Some("c6779ec2960296ed9a04f08d67f64422"), "545d0831-d4ba-408d-b1f1-f4645efb32fd",
```
[the remark this answers is not in the document]
Comment (ruthdas03) [reply test/controllers/SDESCallbackControllerSpec.scala]: Sure, I've split the parameters onto separate lines now.

File: test/controllers/SDESCallbackControllerSpec.scala (hunk 1 above)
Comment (ruthdas03) [reply test/controllers/SDESCallbackControllerSpec.scala]: True. Updated it accordingly.

File: it/controllers/SDESCallbackControllerISpec.scala [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+       |                }
+       |            ]
+       |        }
+       |""".stripMargin
+  )
+
+  "handleCallback" should {
+    "return NO_CONTENT (204)" when {
+      "sdesCallBack is a valid JSON" in {
+        val result: WSResponse = await(buildClientForRequestToApp(uri = "/sdes-callback").post(
```
Comment (LewisMitchell) [inline it/controllers/SDESCallbackControllerISpec.scala:53]: Should we have a test here to check if the audit has been sent?

**Review by LewisMitchell — CHANGES_REQUESTED** [review]

File: it/controllers/SDESCallbackControllerISpec.scala (hunk 8 above)
Comment (ruthdas03) [reply it/controllers/SDESCallbackControllerISpec.scala:53]: Yes, we should. Ameded the same.

**Review by LewisMitchell — APPROVED** [review]

**Pull request merged**